torch/csrc/jit/codegen/cuda/lower_index_compute.cpp

```diff
@@ -715,6 +888,74 @@ IndexFromIdGraph getTensorIndexFromIdGraph(
       loop_indexing.loopDomains());
 }
 
+IndexFromIdGraph getPredicateIndexingFromIdGraph(
+    const std::vector<kir::ForLoop*>& loops,
+    TensorView* consumer_tv,
+    kir::ForLoop* unswitch_or_vec_loop,
+    IterDomain* double_buffer_axis,
+    bool is_start_predicate) {
+  // Run replay pass on the loop nest to generate the deterministic
+  //  traversal info from loop structure.
+  auto loop_indexing =
+      LoopIndexingAnalysis::fromLoopAndConsumer(loops, consumer_tv);
+
+  // Bind initial index variables to the loop nodes and adjust
+  //  according to loop and unswitch info.
+  auto index_parameters = getPredicateInitialIndexParameters(
+      loop_indexing,
+      consumer_tv,
+      unswitch_or_vec_loop,
+      double_buffer_axis,
+      is_start_predicate);
+
+  // Run first backward traversal to generate
+  //  loop nest based indexing math.
+  IndexCompute indexing(
+      index_parameters.initial_concrete_id_index,
+      index_parameters.zero_domains,
+      index_parameters.preferred_concrete_ids,
+      index_parameters.concrete_id_to_halo_extent);
+
+  indexing.run(loop_indexing);
+
+  // Map the concrete id indexing back to consumer tv
+  std::unordered_map<IterDomain*, IterDomain*> index_update_map;
+
+  // First collect all iterdomains in consumer transform history.
+  auto all_consumer_vals = DependencyCheck::getAllValsBetween(
+      {consumer_tv->getMaybeRFactorDomain().begin(),
+       consumer_tv->getMaybeRFactorDomain().end()},
+      {consumer_tv->domain()->domain().begin(),
+       consumer_tv->domain()->domain().end()});
+
+  for (IterDomain* consumer_id :
+       ir_utils::filterByType<IterDomain>(all_consumer_vals)) {
+    // Track the non-concrete id we were trying to bind index
+    //  to, whether from producer or consumer.
+    auto exact_concrete_id = ir_utils::caMapExactConcreteId(consumer_id);
+    index_update_map[exact_concrete_id] = consumer_id;
+  }
+
+  // No contiguity info is used in the predicate indexing pass,
+  //  the predicate generation logic that uses the index math
+  //  generated here will take contiguity into account.
```

Comment:
Yes this PR still not trying to change any semantics yet. Just pure cleaning up. 

I agree it's better to take care of contiguity sooner than later but ideally on the next follow up.

Reply:
I think eventually it'd be most ideal we could just predicate intermediate nodes instead of always going back to root. 

It'd make the generated kernel look cleaner, but doesn't feel like a perf concern to me at the moment.

Reply:
The question isn't what should we do, but what do we do today. I was pretty sure today we treat predicates as fully contiguous and predicate intermediate IterDomains, not just roots. Maybe I'm wrong, maybe @naoyam recalls.

Reply:
I see. This is the section of code on TOT that I was trying to move into `lower_index_compute`:

[https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/index_compute.cpp#L3068-L3087](https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/index_compute.cpp#L3068-L3087)

And we'd still be relying on this part to put in contig info, which isn't refactored yet on this PR:
[https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/index_compute.cpp#L3110-L3111](https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/index_compute.cpp#L3110-L3111)

Reply:
Okay yeah, that's what I thought might be happening. So we are still maintaining same behavior. Will approve.